keps/sig-api-machinery/storage-migration-auto-trigger.md

```diff
@@ -225,6 +208,137 @@ manually, e.g., deploying the [initializer][] manually.
+}
+
+// The storage version hash observed by the reporter at lastHeartbeatTime.
+type StorageVersionHash struct {
+  // The ID of the reporting triggering contoller. This should be the same ID
+  // that's used to contend for a lease, see https://github.com/kubernetes/kubernetes/blob/428a8e04d40ef01c28c66fcfd54f306a1aff9a28/staging/src/k8s.io/api/coordination/v1/types.go#L43.
+  // Required.
+  ReporterID string
+  // The hash value of the current storage version, as shown in the discovery
+  // document served by the API server local to the reporting controller.
+  CurrentStorageVersionHash string
+  // LastHeartbeatTime is the last time the reporting controller checks the
+  // storage version hash of this resource in the discovery document and updates
+  // this field.
+  // +optional
+  LastHeartbeatTime metav1.Time
+}
+``` 
+
+### Extending the work flow of the triggering controller
+
+Normally, HA controllers select a leader, and only the leader writes to the API
+to update status. In this case, we need every triggering controller to report
+its local storage version hash. Specifically, every 10 minutes, each instance of
+the triggering controller 
+* fetches the discovery doc from the local apiserver.
+* adds the storage version hash to PersistedStorageVersionHashes if the local
+  storage version hash is not present there yet.
+* updates (or creates) the element in status.storageVersionhashes with the
+  matching reporterID, recording the latest currentStorageVersionHash and
+  lastHeartbeatTime.
+* removes storageVersionHashes whose lastHeartbeatTime are more than 2 polling
+  periods (20 mins) old. This garbage collection is necessary because the
+  reporter ID [changes][] every time a process reboots. The garbage collection
+  happens 20 mins after lastHeartbeat to give the reporter a chance to recover
+  from transient errors.
+
+[changes]:https://github.com/kubernetes/kubernetes/blob/428a8e04d40ef01c28c66fcfd54f306a1aff9a28/cmd/kube-controller-manager/app/controllermanager.go#L238
+
+Only the leading triggering controller reacts to the storageState changes. The
+leading triggering controller
+* deletes pending storageVersionMigration objects to cancel any ongoing
+  migration, once the hashes in currentStorageVersionHashes stop being
+  identical.
+* creates a storageVersionMigration object to start migration, if
+  * persistedStorageVersionHashes has "Unknown" or has more than one version,
+  * **AND** the lastHeartbeatTime in all items of the CurrentStorageVersionHashes
+  are no older than 1 polling period (10 mins)
+  * **AND** the hashes in currentStorageVersionHashes have been identical in the
+    past two polls. In other words, the hashes have been held identical for at
+    least one polling period (10 minutes). The purpose is to give newly deployed
+    triggering controller a chance to report the hash.
+
+### Case walkthrough
+
+Case 1. Assuming the HA cluster consists of 3 apiservers. The cluster admin
```

Comment:
one interesting case is: normally the HA cluster has three apiservers, but for a prolonged period of time, one apisever keeps crashing, so essentially only two apiservers are working.

Reply:
Yeah, that's worth documenting. I'll add it if we decide to pursue this design.